motionplan/rrtStarConnect.go

```diff
@@ -176,64 +235,208 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
+			if nearDist < math.Pow(mp.algOpts.JointSolveDist, 3) {
+				if !doubled {
+					doubled = true
+					// Check if doubling qstep will allow escape from the identical configuration
+					// If not, we terminate and return.
+					// If so, qstep will be reset to its original value after the rescue.
+					for i, q := range qstep {
+						qstep[i] = q * 2.0
+					}
+					continue
+				} else {
+					// We've arrived back at very nearly the same configuration again; stop solving and send back oldNear.
+					// Do not add the near-identical configuration to the RRT map
+					mchan <- oldNear
+					return
+				}
+			}
+			if doubled {
+				copy(qstep, mp.algOpts.qstep)
+				doubled = false
+			}
+			// constrainNear will ensure path between oldNear and newNear satisfies constraints along the way
+			near = &basicNode{q: newNear, cost: neighbors[0].node.Cost() + neighbors[0].dist}
+			rrtMap[near] = oldNear
+
+			// rewire the tree
+			for i, thisNeighbor := range neighbors {
+				// dont need to try to rewire nearest neighbor, so skip it
+				if i == 0 {
+					continue
+				}
+
+				// check to see if a shortcut is possible, and rewire the node if it is
+				connectionCost := mp.planOpts.DistanceFunc(&Segment{
+					StartConfiguration: thisNeighbor.node.Q(),
+					EndConfiguration:   near.Q(),
+				})
+				cost := connectionCost + near.Cost()
+
+				// If 1) we have a lower cost, and 2) the putative updated path is valid
+				if cost < thisNeighbor.node.Cost() && mp.checkPath(target.Q(), thisNeighbor.node.Q()) {
+					// Alter the cost of the node
+					// This needs to edit the existing node, rather than make a new one, as there are pointers in the tree
+					thisNeighbor.node.SetCost(cost)
+					rrtMap[thisNeighbor.node] = near
+				}
+			}
+		} else {
 			break
 		}
 	}
+	mchan <- oldNear
+}
 
-	// add new node to tree as a child of the minimum cost neighbor node if it was reachable
-	if minIndex == -1 {
-		mchan <- nil
-		return
-	}
-	targetNode := &basicNode{q: target, cost: minCost}
-	tree[targetNode] = neighbors[minIndex].node
-
-	// rewire the tree
-	for i, thisNeighbor := range neighbors {
-		// dont need to try to rewire minIndex, so skip it
-		if i == minIndex {
-			continue
-		}
-
-		// check to see if a shortcut is possible, and rewire the node if it is
-		connectionCost := mp.planOpts.DistanceFunc(&Segment{
-			StartConfiguration: thisNeighbor.node.Q(),
-			EndConfiguration:   targetNode.Q(),
-		})
-		cost := connectionCost + targetNode.Cost()
-
-		// If 1) we have a lower cost, and 2) the putative updated path is valid
-		if cost < thisNeighbor.node.Cost() && mp.checkPath(target, thisNeighbor.node.Q()) {
-			// Alter the cost of the node
-			// This needs to edit the existing node, rather than make a new one, as there are pointers in the tree
-			thisNeighbor.node.SetCost(cost)
-			tree[thisNeighbor.node] = targetNode
+func (mp *rrtStarConnectMotionPlanner) constrainNear(
```

Comment:
I'm not convinced we want to add constraint satisfaction to RRT* at this time. If not, this whole function can likely be replaced with a single `mp.planOpts.CheckSegmentAndStateValidity` call

Reply:
I agree that makes sense.